# Decision Tree : 이진 분류 실습

## 데이터 로드

In [1]:
import pandas as pd

df = pd.read_csv('mushrooms.csv')
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color   ...   stalk-surface-below-ring  \
0            c         n          k   ...                          s   
1            c         b          k   ...                          s   
2            c         b          n   ...                          s   
3            c         n          n   ...                          s   
4            w         b          k   ...                          s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [2]:
df.shape

(8124, 23)

In [3]:
df.info()
#null 값이 없네. 전처리가 돼있네!

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
class                       8124 non-null object
cap-shape                   8124 non-null object
cap-surface                 8124 non-null object
cap-color                   8124 non-null object
bruises                     8124 non-null object
odor                        8124 non-null object
gill-attachment             8124 non-null object
gill-spacing                8124 non-null object
gill-size                   8124 non-null object
gill-color                  8124 non-null object
stalk-shape                 8124 non-null object
stalk-root                  8124 non-null object
stalk-surface-above-ring    8124 non-null object
stalk-surface-below-ring    8124 non-null object
stalk-color-above-ring      8124 non-null object
stalk-color-below-ring      8124 non-null object
veil-type                   8124 non-null object
veil-color                  8124 non-null object
ring-number

##  Train / Test set 나누기
### 1. 고전적인 방법
* df_X = df.drop(columns=['class'])
* df_y = df['class']
* X_train = df_x.iloc[0:270] 
* Y_train = df_y.iloc[0:270]  이런 식으로 분할할 수 있음

### 2. sklearn 라이브러리 활용
* from sklearn.model_selection import train_test_split
* df_X = df.drop(columns=['class'])
* df_y = df['class']
* X_train, X_test = train_test_split(X, test_size=0.3, random_state=42)
* y_train, y_test = train_test_split(y, test_size=0.3, random_state=42)
* 또는,  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

* type(X_train) : pandas.core.frame.DataFrame
* type(y_tain) : pandas.core.series.Series -> 1차원 즉, (x, ) 형태

## 오류 : Decision Tree 에서는 string(X)
* No! 오류 발생 : could not convert string to float: 'g' 

* 범주형 변수 > 수치형 변수 변환
 - LavelEncoder
 
* 인코딩은 카테고리 값이나 텍스트 정보를 처리가 쉬운 정수로 변환하는 과정이다.


In [4]:
df_X = df.drop(columns=['class'])
df_y = df['class']

### LabelEncoder

In [5]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
le.fit(df_y)
le.classes_
df_y=le.fit_transform(df_y)
# le.inverse_transform(y2)

In [6]:
df_y

array([1, 0, 0, ..., 0, 1, 0], dtype=int64)

* le.fit(df_X)
* error : bad input shape (8124, 22)
* 한 컬럼씩 적용해야함. 따라서 for문 돌리기

In [7]:
df_X.columns

Index(['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'],
      dtype='object')

In [8]:
for i in df_X.columns:
    df_X[i] = le.fit_transform(df_X[i])

In [9]:
df_X.head()

cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0          5            2          4        1     6                1   
1          5            2          9        1     0                1   
2          0            2          8        1     3                1   
3          5            3          8        1     6                1   
4          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  stalk-shape   ...     \
0             0          1           4            0   ...      
1             0          0           4            0   ...      
2             0          0           5            0   ...      
3             0          1           5            0   ...      
4             1          0           4            1   ...      

   stalk-surface-below-ring  stalk-color-above-ring  stalk-color-below-ring  \
0                         2                       7                       7   
1                         2                       7                       7   
2                         2                       7                       7   
3                         2                       7                       7   
4                         2                       7                       7   

   veil-type  veil-color  ring-number  ring-type  spore-print-color  \
0          0           2            1          4                  2   
1          0           2            1          4                  3   
2          0           2            1          4                  3   
3          0           2            1          4                  2   
4          0           2            1          0                  3   

   population  habitat  
0           3        5  
1           2        1  
2           2        3  
3           3        5  
4           0        1  

[5 rows x 22 columns]

## Train / Test 분할

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.3, random_state=42)

In [11]:
X_train.head(2)

cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
5921          5            2          0        1     2                1   
1073          5            0          3        1     5                1   

      gill-spacing  gill-size  gill-color  stalk-shape   ...     \
5921             0          0           3            1   ...      
1073             0          0           7            1   ...      

      stalk-surface-below-ring  stalk-color-above-ring  \
5921                         0                       7   
1073                         2                       6   

      stalk-color-below-ring  veil-type  veil-color  ring-number  ring-type  \
5921                       7          0           2            1          4   
1073                       7          0           2            1          4   

      spore-print-color  population  habitat  
5921                  1           4        5  
1073                  2           5        0  

[2 rows x 22 columns]

In [12]:
y_train[0:10]

array([1, 0, 1, 0, 1, 0, 1, 1, 0, 1], dtype=int64)

## sklearn 라이브러리의 tree 활용

In [13]:
from sklearn import tree
clf = tree.DecisionTreeClassifier(criterion='entropy')
# clf = tree.DecisionTreeClassifier()


# 순도 계산법 : 1. 엔트로피 2. 지니
# max_depth
clf.fit(X_train, y_train )

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [14]:
clf.predict_proba(X_test)
#Predict class probabilities of the input samples X

array([[1., 0.],
       [0., 1.],
       [0., 1.],
       ...,
       [0., 1.],
       [1., 0.],
       [1., 0.]])

In [15]:
pre = clf.predict(X_test)
#Predict class or regression value for X.
pre

array([0, 1, 1, ..., 1, 0, 0], dtype=int64)

In [16]:
y_test

array([0, 1, 1, ..., 1, 0, 0], dtype=int64)

## accuracy

In [17]:
from sklearn.metrics import accuracy_score

y_pre = clf.predict(X_test)
print('Accuracy: %.2f' % accuracy_score(y_test, y_pre))

Accuracy: 1.00


In [18]:
print((pre == y_test).mean())

1.0


## 시각화

* No module named 'graphviz' -> 아나콘타 프롬프트 : pip install grapviz

In [19]:
X_test.head(2)

cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
1971          2            0          4        0     5                1   
6654          2            2          2        0     8                1   

      gill-spacing  gill-size  gill-color  stalk-shape   ...     \
1971             1          0           3            1   ...      
6654             0          1           0            1   ...      

      stalk-surface-below-ring  stalk-color-above-ring  \
1971                         0                       7   
6654                         2                       6   

      stalk-color-below-ring  veil-type  veil-color  ring-number  ring-type  \
1971                       7          0           2            1          0   
6654                       6          0           2            1          0   

      spore-print-color  population  habitat  
1971                  3           3        1  
6654                  7           4        2  

[2 rows x 22 columns]

In [20]:
df_y

array([1, 0, 0, ..., 0, 1, 0], dtype=int64)

* 결국 GraphViz's executables not found 문제를 해결하지 못함
* 시각화 실패 

In [23]:
# from sklearn.tree import export_graphviz
# import pydotplus
# from IPython.display import Image

# dot_data = export_graphviz(clf, out_file=None, feature_names=df_X.columns,
#                            class_names=['poision','e'],
#                            filled=True, rounded=True,
#                            special_characters=True)
# graph = pydotplus.graph_from_dot_data(dot_data)
# Image(graph.create_png())

# Random Forest : 2진 분류

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
RF = RandomForestClassifier(n_estimators=100, max_depth=7, random_state=0)
RF.fit(X_train, y_train)

In [ ]:
pre_R = RF.predict(X_test)

In [ ]:
print((pre_R == y_test).mean())

* default 값이 더 정확함

In [ ]:
RF_2 = RandomForestClassifier()
RF_2.fit(X_train, y_train)

In [ ]:
pre_R2 = RF_2.predict(X_test)
print((pre_R2 == y_test).mean())


## 모델을 적용한 결과 활용

  1. 예측값이 array 형태로 돼 있음 
  2. string 형태변환 한 뒤 DataFrame에 합쳐도 되는 거고.
  3. 의문점 : label 값이 없는 데이터들을 분류하려면?
      - 예를 들어, 카카오톡 아레나 카테고리 분류 문제
      - 비지도학습 머신러닝 알고리즘을 사용하면 되겠네!

In [ ]:
pre_R2